src/js/tracing/integrations/nativeFrames.ts

```diff
@@ -1,239 +1,201 @@
-    if (isRootSpan(span)) {
-      return _onTransactionFinish(span);
-    }
-
+  const fetchEndFramesForSpan = (span: Span): void => {
     const timestamp = timestampInSeconds();
+    const spanId = span.spanContext().spanId;
 
-    void NATIVE.fetchNativeFrames()
-      .then(frames => {
-        if (!frames) {
-          return;
-        }
-
-        _lastSpanFinishFrames = {
-          timestamp,
-          nativeFrames: frames,
-        };
-      })
-      .then(undefined, error => {
-        logger.error(`[NativeFrames] Error while fetching frames for child span end.`, error);
-      });
-  };
+    if (isRootSpan(span)) {
+      const hasStartFrames = _spanToNativeFramesAtStartMap.has(spanId);
+      if (!hasStartFrames) {
+        // We don't have start frames, won't be able to calculate the difference.
+        return;
+      }
 
-  /**
-   * To be called when a transaction is finished
-   */
-  const _onTransactionFinish = (span: Span): void => {
-    _fetchFramesForTransaction(span).then(undefined, (reason: unknown) => {
-      logger.error(
-        `[NativeFrames] Error while fetching frames for root span start (${span.spanContext().spanId})`,
-        reason,
+      logger.debug(`[${INTEGRATION_NAME}] Fetch frames for root span end (${spanId}).`);
+      _spanToNativeFramesAtEndMap.set(
+        spanId,
+        new Promise<NativeFramesResponseWithTimestamp | null>(resolve => {
+          fetchNativeFrames()
+            .then(frames => {
+              resolve({
+                timestamp,
+                nativeFrames: frames,
+              });
+            })
+            .then(undefined, error => {
+              logger.debug(`[${INTEGRATION_NAME}] Error while fetching native frames.`, error);
+              resolve(null);
+            });
+        }),
       );
-    });
+      return undefined;
+    } else {
+      logger.debug(`[${INTEGRATION_NAME}] Fetch frames for child span end (${spanId}).`);
+      fetchNativeFrames()
+        .then(frames => {
+          _lastChildSpanEndFrames = {
+            timestamp,
+            nativeFrames: frames,
+          };
+        })
+        .catch(error => logger.debug(`[${INTEGRATION_NAME}] Error while fetching native frames.`, error));
+    }
   };
 
-  /**
-   * Returns the computed frames measurements and awaits for them if they are not ready yet.
-   */
-  const _getFramesMeasurements = (
-    traceId: string,
-    finalEndTimestamp: number,
-    startFrames: NativeFramesResponse,
-  ): Promise<FramesMeasurements | null> => {
-    if (_finishFrames.has(traceId)) {
-      logger.debug(`[NativeFrames] Native end frames already fetched for trace id (${traceId}).`);
-      return Promise.resolve(_prepareMeasurements(traceId, finalEndTimestamp, startFrames));
+  const processEvent = async (event: Event): Promise<Event> => {
+    if (
+      event.type !== 'transaction' ||
+      !event.transaction ||
+      !event.contexts ||
+      !event.contexts.trace ||
```

Comment:
We could check if contexts exits by checking trace itself.

```suggestion
      !event.contexts?.trace ||
```